plutus-ledger-api/src/PlutusLedgerApi/Common/Eval.hs

```diff
@@ -106,33 +106,71 @@ mkTermToEvaluate ll pv script args = do
     through (liftEither . first DeBruijnError . UPLC.checkScope) appliedT
 
 toMachineParameters :: MajorProtocolVersion -> EvaluationContext -> DefaultMachineParameters
-toMachineParameters _ = machineParameters
+toMachineParameters pv (EvaluationContext lv toSemVar machParsList) =
+    case lookup (toSemVar pv) machParsList of
+        Nothing -> error $ Prelude.concat
+            ["Internal error: ", show lv, " does not support protocol version ", show pv]
+        Just machPars -> machPars
 
 {-| An opaque type that contains all the static parameters that the evaluator needs to evaluate a
-script.  This is so that they can be computed once and cached, rather than being recomputed on every
+script. This is so that they can be computed once and cached, rather than being recomputed on every
 evaluation.
+
+Different protocol versions may require different bundles of machine parameters, which allows us for
+example to tweak the shape of the costing function of a builtin, so that the builtin costs less.
+Currently this means that we have to create multiple 'DefaultMachineParameters' per language
+version, which we put into a cache (represented by an association list) in order to avoid costly
+recomputation of machine parameters.
+
+In order to get the appropriate 'DefaultMachineParameters' at validation time we look it up in the
+cache using a semantics variant as a key. We compute the semantics variant from the protocol
+version using the stored function. Note that the semantics variant depends on the language version
+too, but the latter is known statically (because each language version has its own evaluation
+context), hence there's no reason to require it to be provided at runtime.
```

Comment:
So in principle we have a matrix of semantics variants index by (LL,PV) pairs, but at this point we know the LL, so `machParsList` is essentially the row of that matrix corresponding to the given LL.  Is that a reasonable way to think of this?

Reply:
> So in principle we have a matrix of semantics variants index by (LL,PV) pairs, but at this point we know the LL, so `machParsList` is essentially the row of that matrix corresponding to the given LL.  Is that a reasonable way to think about this?

Yes, exactly. I'll steal your wording for the comment I think. Thanks!